src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreRT.cs

```diff
@@ -417,14 +405,68 @@ public static unsafe IntPtr StringToBSTR(string s)
 
         public static void WriteByte(object ptr, int ofs, byte val)
         {
-            // Obsolete
-            throw new PlatformNotSupportedException("WriteByte");
+            WriteValueSlow(ptr, ofs, val, (IntPtr nativeHome, int offset, byte value) => WriteByte(nativeHome, offset, value));
+        }
+
+        public static void WriteInt16(object ptr, int ofs, short val)
+        {
+            WriteValueSlow(ptr, ofs, val, WriteInt16);
         }
 
         public static void WriteInt32(object ptr, int ofs, int val)
         {
-            // Obsolete
-            throw new PlatformNotSupportedException("WriteInt32");
+            WriteValueSlow(ptr, ofs, val, WriteInt32);
+        }
+
+        public static void WriteInt64(object ptr, int ofs, long val)
+        {
+            WriteValueSlow(ptr, ofs, val, WriteInt64);
+        }
+
+        private static unsafe void WriteValueSlow<T>(object ptr, int ofs, T val, Action<IntPtr, int, T> writeValueHelper)
+        {
+            // Consumers of this method are documented to throw AccessViolationException on any AV
+            if (ptr is null)
+            {
+                throw new AccessViolationException();
+            }
+
+            if (ptr.EETypePtr.IsArray ||
+                ptr is string ||
+                ptr is StringBuilder)
+            {
+                // We could implement these if really needed.
+                throw new PlatformNotSupportedException();
+            }
+
+            // We are going to assume this is a Sequential or Explicit layout type because
+            // we don't want to touch reflection metadata for this.
+            // If we're wrong, this will throw the exception we get for missing interop data
+            // instead of an ArgumentException.
+            // That's quite acceptable for an obsoleted API.
+
+            Type structType = ptr.GetType();
+
+            int size = SizeOf(structType);
+
+            // Compat note: CLR wouldn't bother with a range check. If someone does this,
+            // they're likely taking dependency on some CLR implementation detail quirk.
+            if (checked(ofs + Unsafe.SizeOf<T>()) > size)
+                throw new ArgumentOutOfRangeException(nameof(ofs));
+
+            IntPtr nativeBytes = AllocCoTaskMem(size);
```

Comment:
Clear the memory so that `DestroyStructure` does not see random data when `StructureToPtr` throws?

(Same for Read)